Add Payment Channel and Method to Curb Events

jiffyclub opened the pull request:
## Explain pull request

Cities are often interested in how their curb users are paying for their time at the curb, for example the breakdown between cash, cards, mobile, etc. This adds a `payment_channel` and `payment_method` fields to the Curb Event model and includes a list of potential values for each.

The "channel" describes the medium by which a payment was made, such as at a physical meter or via a mobile app. The "method" describes how money was exchanged, such as via cash or a credit card.

Closes #157.

## Is this a breaking change

* No, not breaking

## Impacted Spec

Which API(s) will this pull request impact?

* `Events`

## Additional context

I totally just made up many of the items in the suggested list of payment types, input from practitioners on changes or additions is welcome.

jiffyclub committed:
Add Payment Type to Curb Events

Cities are often interested in how their curb users are paying for their
time at the curb, for example the breakdown between cash, cards, mobile, etc.
This adds a payment_type field to the Curb Event model and includes a list
of potential payment types.

Closes #157.

schnuerle commented on events/README.md:
```diff
@@ -247,6 +248,24 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| `cash`            | Bills or coins at a meter. |
+| `credit_card`     | Visa, Mastercard, etc at a meter. |
+| `membership_card` | A card used at a meter to pay via a corporate membership or loyalta program, etc. |
+| `mobile`          | Paid via a mobile app. |
```
Could add QR code or app clip type payment where it's on mobile but not an app.

schnuerle commented on events/README.md:
```diff
@@ -247,6 +248,24 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| ----------------- | ------------------------------------------------------ |
+| `cash`            | Bills or coins at a meter. |
+| `credit_card`     | Visa, Mastercard, etc at a meter. |
+| `membership_card` | A card used at a meter to pay via a corporate membership or loyalta program, etc. |
```
Could also be smart card or other type of card.

hdmm3 replied:
Small typo: `loyalty`

schnuerle commented on events/README.md:
```diff
@@ -247,6 +248,24 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| Name              | Description                                            |
+| ----------------- | ------------------------------------------------------ |
+| `cash`            | Bills or coins at a meter. |
+| `credit_card`     | Visa, Mastercard, etc at a meter. |
```
Idea to add sub payment type as an option under `payment_type`. Visa, MC, apple pay, etc. Or V2X payment possible.

michael-danko-passport replied:
I'm not sure there's enough distinction to justify separating mobile_app and mobile_app_clip. Both iOS and Android treat these lightweight versions (like App Clips for iOS or Instant Apps for Android) as part of the same application from the user's perspective.

On the implementation side, they typically interact with the same backend APIs and follow similar transaction flows. As a result, it may be more practical to group them under a single mobile_app designation unless there’s a specific use case that requires tracking them separately.

To consolidate under a single mobile_app, we could use the definition: Paid via a mobile app, including iOS App Clips, and Android Instant App.

schnuerle replied:
Using just mobile_app now, with updated description.

schnuerle commented on events/README.md:
```diff
@@ -247,6 +248,24 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| `sms`             | Paid via text message. |
+| `billing`         | Curb user will be billed for usage at a later time. |
+| `permit`          | Curb user has a permit allowing them to use the curb without payment. |
+| `courtesy`        | At a curb that normally requires payment this event for some reason did not. |
```
`voucher` and `test` might also be useful options from MDS.
https://github.com/openmobilityfoundation/mobility-data-specification/blob/main/modes/passenger-services.md#fare-attributes

hdmm3 commented on events/README.md:
```diff
@@ -247,6 +248,24 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| `cash`            | Bills or coins at a meter. |
+| `credit_card`     | Visa, Mastercard, etc at a meter. |
+| `membership_card` | A card used at a meter to pay via a corporate membership or loyalta program, etc. |
+| `mobile`          | Paid via a mobile app. |
```
`mobile` could potentially be ambiguous, since payment with a mobile app could be done with credit/debit cards, but also with digital wallets, prepaid accounts, permits, etc.

schnuerle replied:
`mobile_app` has been suggested above.

hdmm3 commented:
I wonder if it's worth distinguishing between the _payment channel_ and the _payment method_. 
The payment channel is the medium or platform used to pay: meter, mobile app, website, sms. The payment method would be the instrument used to pay: cash, credit/debit card, transit card, prepaid account, permit, etc.

jiffyclub committed:
Split payment_type into channel and method enums

Split the payment_type field into payment_channel and payment_type
so that we can describe both the channel via which payment was made
(at a meter, via mobile app, etc) from how money was transferred
(via cash, credit card, etc).

jiffyclub commented:
I've split the `payment_type` into `payment_channel` and `payment_method` enums.

I'm not sure where the `billing` item should go... maybe both places?

jacobmalleau commented:
I agree with what is being proposed here, definitely something CurbIQ would use.

mplsmitch commented on events/README.md:
```diff
@@ -247,6 +249,41 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| `cash`            | Bills or coins at a meter. |
+| `credit_card`     | Visa, Mastercard, etc at a meter. |
+| `smart_card`      | A specialized smart card. |
+| `membership_card` | A card used at a meter to pay via a corporate membership or loyalty program, etc. |
```
`mobile_wallet` for example: Apple Pay, Google Pay, Samsung Pay, PayPal, Zelle, Venmo

michael-danko-passport replied:
I think `digital_wallet` can also include an app specific or 'closed-loop' wallet. Not sure if that's worth adding to the example description.

mplsmitch commented:
We will be discussing this at the Feb 25th [CDS Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.02.25-Curb). Please review and leave your comments and questions ahead of time.

schnuerle commented:
Comments from working group meeting 

- Matt Davis (MD) of  Populus - suggestion to split the way user interacted vs how they paid for curb. PR adds payment method and payment channel with method being things like credit card or cash and channel being things like meter or mobile app.
- Mitch V - everyone go and look at PR and add any enumerables that are missing from the two fields.
- MV- are there other things we are not covering, connected vehicles bluetooth, RFID etc

schnuerle commented:
Do we need a 'transaction_id' and `payment_vendor_id` (UUID from [data source file](https://github.com/openmobilityfoundation/curb-data-specification/blob/main/data_source_operators.csv)) field as well to connect this to external payment platforms.

rneubauer commented:
I think this is important and have talked with a few city customers about it. One thing I also think is important is the transaction ID. This will provide a way to match up the transaction you get in a CDS event to the actual transaction that is in the third party vendor so there is ability to do a full circle lookup.

rneubauer commented:
I second the Transaction ID for sure.

schnuerle commented:
To keep this in 1.1 we are going to have to resolve the conflicts, and incorporate feedback into the PR.

jiffyclub committed:
Add payment_transaction_id field

jiffyclub committed:
Add digital_wallet payment type.

jiffyclub commented:
I've rebased on `dev` and added a `payment_transaction_id` field to the Event schema, and added a `digital_wallet` option to the payment type enum.

schnuerle commented on events/README.md:
```diff
@@ -253,6 +256,42 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
+| ----------------- | ------------------------------------------------------ |
+| `meter`           | User paid at a physical meter. |
+| `mobile_app`      | Paid via a mobile app. |
+| `mobile_app_clip` | Paid via a temporarily downloaded mobile app clip, not a full mobile app. |
```
As stated below by @michael-danko-passport do we really need this option? Description for `mobile_app` could read " Paid via a mobile app, including iOS App Clips, and Android Instant App."

JohnLundstrom-Minneapolis replied:
I often will see companies market a third payment option (apart from the terminal or an app) as "web" if we wished to differentiate it from app and meter payments.

schnuerle replied:
We added the combined these so mobile_app handles both, and added website.

schnuerle commented:
Hi @jiffyclub could you resolve the new conflicts with the dev branch. It's not allowing me to. There have been updates from other PRs.

schnuerle commented:
We will be reviewing this at [tomorrow's public Curb Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.08.12-Curb). However we won't have much time to discuss, so leave your comments and suggestions here as general or inline comments to spark changes before the [CDS 1.1 release](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Release-1.1.0) is made.

jiffyclub commented:
Rebased to fix conflicts. I don't have strong opinions on the different payment methods people are proposing so I'd like to get to something we can merge and then if people want to make edits they can make their own PRs.

schnuerle commented on events/README.md:
```diff
@@ -309,6 +312,42 @@ Type(s) of lane used or blocked `vehicle_blocked_lane_types` by the vehicle perf
 [Top][toc]
 
+
+### Payment Channel
```
Can you add this and the payment method sections to the TOC at the top?

jiffyclub committed:
Add new enums to TOC

jiffyclub committed:
Remove "mobile app clip" payment channel.

jiffyclub committed:
Add telephone payment channel

schnuerle commented:
@jiffyclub Can you turn on admin editing on this PR? It might be too late for that.

If so, can you make these changes based on the WG discussion?

- Add new channel value, `phone` for when people can call a phone number to pay via automated phone tree or talking to a person, etc. 

I think that's all that came out of the discussions, but if you picked up on other things, please tweak the PR.

jiffyclub commented:
I added the `telephone` payment type in https://github.com/openmobilityfoundation/curb-data-specification/pull/158/commits/a0f37e874f5752fd465898b071e368fcaae75ebf
